batch/utils/copy_files.py

```diff
@@ -0,0 +1,396 @@
+            size = self.length
+        size = min(size, self._end - self._fp.tell())
+        return self._fp.read(max(0, size))
+
+    def seek(self, offset, whence=os.SEEK_SET):
+        if whence == os.SEEK_END:
+            return self._fp.seek(offset + self._end)
+        elif whence == os.SEEK_CUR:
+            return self._fp.seek(offset, whence)
+        else:
+            assert whence == os.SEEK_SET
+            return self._fp.seek(self._start + offset)
+
+    def close(self):
+        self._fp.close()
+
+    def flush(self, size=None):
+        raise NotImplementedError('flush is not implemented in FilePart.')
+
+    def fileno(self, size=None):
+        raise NotImplementedError('fileno is not implemented in FilePart.')
+
+    def isatty(self, size=None):
+        raise NotImplementedError('isatty is not implemented in FilePart.')
+
+    def next(self, size=None):
+        raise NotImplementedError('next is not implemented in FilePart.')
+
+    def readline(self, size=None):
+        raise NotImplementedError('readline is not implemented in FilePart.')
+
+    def readlines(self, size=None):
+        raise NotImplementedError('readlines is not implemented in FilePart.')
+
+    def xreadlines(self, size=None):
+        raise NotImplementedError('xreadlines is not implemented in FilePart.')
+
+    def truncate(self, size=None):
+        raise NotImplementedError('truncate is not implemented in FilePart.')
+
+    def write(self, size=None):
+        raise NotImplementedError('write is not implemented in FilePart.')
+
+    def writelines(self, size=None):
+        raise NotImplementedError('writelines is not implemented in FilePart.')
+
+
+def is_gcs_path(file):
+    return file.startswith('gs://')
+
+
+def flatten(its):
+    return [x for it in its for x in it]
+
+
+def listdir(path):
+    if not os.path.exists(path):
+        raise FileNotFoundError(path)
+    if os.path.isfile(path):
+        return [(path, os.path.getsize(path))]
+    # gsutil doesn't copy empty directories
+    return flatten([listdir(path.rstrip('/') + '/' + f) for f in os.listdir(path)])
+
+
+def get_dest_path(file, src, include_recurse_dir):
+    src = src.rstrip('/').split('/')
+    file = file.split('/')
+    if len(src) == len(file):
+        return file[-1]
+
+    # https://cloud.google.com/storage/docs/gsutil/commands/cp#how-names-are-constructed_1
+    if include_recurse_dir:
+        recurse_point = len(src) - 1
+    else:
+        recurse_point = len(src)
+
+    return '/'.join(file[recurse_point:])
+
+
+def get_partition_starts(file_size, max_partitions, min_partition_size):
+    if file_size == 0:
+        return [0, 0]
+
+    if file_size / min_partition_size > max_partitions:
+        partition_size = (file_size + max_partitions - 1) // max_partitions
+    else:
+        partition_size = min_partition_size
+
+    partition_starts = [i for i in range(0, file_size, partition_size)]
+    partition_starts.append(file_size)
+    return partition_starts
```

Comment:
What is the goal of this function?

I think the goal for this function is to:
- create evenly balanced partitions
given two constraints:
- the partitions can be no smaller than `min_partition_size`, and
- the number of partitions can be no more than `max_partitions`.

`get_partition_starts` does not always produce evenly balanced partitions. For example:

```
In [9]: get_partition_starts(33000, 10, 10000)                                                                                                                                                                                                                         
Out[9]: [0, 10000, 20000, 30000, 33000]
```

```
boundaries = partition_boundaries(22, 10, 7)
for x, y in zip(boundaries, boundaries[1:]):
    assert (y - x) == 7 or (y - x) == 8
```

Perfect balance is given by this many partitions: `file_size / min_partition_size`. Unfortunately, we can't have fractional partitions, so we need to redistribute the fraction evenly among all the other partitions. In general, we should be able to create a partitioning where every partition is size `x` or `x+1` for some `x` which is not much bigger or equal to the minimum partition size.

Reply:
erm I thought about this more and maybe I am wrong. We probably want to align to some large power of two, right? On the other hand:
```
parallelism=16 block_size=2Gi replica=1
copying gs://hail-jigold-59hi5/data/large to /tmp/large of size 5.4 GB...
copied gs://hail-jigold-59hi5/data/large to /tmp/large of size 5.4 GB in 73.782s
  setup: 1.811s
  download 4294967296-5368709119: 48.524s
  download 0-2147483647: 71.478s
  download 2147483648-4294967295: 71.965s
```
This benchmark shows  the tiny partition is significantly faste.r

Reply:
I made the partitions as balanced as possible